### ISSUE TITLE
Collaborative Based Recommendation System

### ISSUE THREAD

**Issue by talhahashim** [talhahashim/Collaborative-Based-Recommendation-System#1]
import pandas as pd
import numpy as np
from surprise import Dataset, Reader, SVD
from surprise.model_selection import train_test_split
from surprise import accuracy
from collections import defaultdict
import os

# === 1. Parse Netflix Prize Data (Kaggle format) ===
def load_combined_data(folder_path='.', files=None):
    """
    Reads combined_data_X.txt files and returns a DataFrame with
    columns: userId (CustomerID), movieId, rating, date.
    """
    if files is None:
        files = [f for f in os.listdir(folder_path) if f.startswith('combined_data_') and f.endswith('.txt')]
    data = []
    for fname in sorted(files):
        print(f"Processing {fname}...")
        current_movie = None
        with open(os.path.join(folder_path, fname), 'r') as f:
            for line in f:
                line = line.strip()
                if line.endswith(':'):
                    current_movie = int(line[:-1])
                else:
                    user, rating, date = line.split(',')
                    data.append((int(user), current_movie, float(rating), date))
    return pd.DataFrame(data, columns=['userId', 'movieId', 'rating', 'date'])

# Example: Adjust folder_path if needed
df = load_combined_data(folder_path='.', files=['combined_data_1.txt'])  # maybe start with first file for speed

# Optional sampling for a subset to fit memory constraints
# df = df.sample(n=5000000, random_state=42)

print("Total ratings loaded:", len(df))
print(df.head())

# === 2. Prepare Data for Surprise SVD ===
from surprise import Reader, Dataset
reader = Reader(rating_scale=(1, 5))
data = Dataset.load_from_df(df[['userId', 'movieId', 'rating']], reader)

# === 3. Train-Test Split ===
trainset, testset = train_test_split(data, test_size=0.2, random_state=42)

# === 4. Build and Train SVD Model ===
model = SVD(n_factors=100, reg_all=0.02, lr_all=0.005, random_state=42)
model.fit(trainset)

# === 5. Predict and Evaluate ===
predictions = model.test(testset)
rmse = accuracy.rmse(predictions)
print(f"Test RMSE: {rmse:.5f}")

# === 6. Generate Top-N Recommendations per User ===
def get_top_n(predictions, n=5):
    # Map predictions to per-user lists
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))
    # Sort and select top-n
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]
    return top_n

top_n = get_top_n(predictions, n=5)

# Display sample recommendations
print("\nSample Top-5 Recommendations:")
for uid, user_ratings in list(top_n.items())[:5]:
    print(f"User {uid}:")
    for movie_id, est in user_ratings:
        print(f"  Movie {movie_id} — Predicted Rating: {est:.2f}")
